nucypher/blockchain/eth/clients.py

```diff
@@ -38,97 +38,121 @@ class Web3ClientUnexpectedVersionString(Web3ClientError):
+        Parity  -> 'Parity-Ethereum/v2.5.1-beta-e0141f8-20190510/x86_64-linux-gnu/rustc1.34.1'
+        Ganache -> 'EthereumJS TestRPC/v2.1.5/ethereum-js'
+        PyEVM   -> 'EthereumTester/0.1.0b39/linux/python3.6.7'
+        """
+        clients = {
+
+            # Geth
+            cls.GETH: GethClient,
+
+            # Parity
+            cls.PARITY: ParityClient,
+            cls.ALT_PARITY: ParityClient,
+
+            # Test Clients
+            cls.GANACHE: GanacheClient,
+            cls.ETHEREUM_TESTER: EthereumTesterClient,
+        }
 
         try:
-            subcls = {
-                GETH: GethClient,
-                PARITY: ParityClient,
-                ALT_PARITY: ParityClient,
-                GANACHE: GanacheClient,
-                ETHEREUM_TESTER: EthTestClient,
-            }[node_technology]
+            client_data = w3.clientVersion.split('/')
+            node_technology = client_data[0]
+            ClientSubclass = clients[node_technology]
+
+        except (ValueError, IndexError):
+            raise ValueError(f"Invalid client version string. Got '{w3.clientVersion}'")
+
         except KeyError:
-            raise NotImplementedError(node_technology)
+            raise NotImplementedError(f'{w3.clientVersion} is not a supported ethereum client')
+
+        client_kwargs = {
+            'node_technology': node_technology,
+            'version': client_data[1],
+            'backend': client_data[-1],
+            'platform': client_data[2] if len(client_data) == 4 else None  # Plaftorm is optional
+        }
 
-        return subcls(w3, *client_data)
+        instance = ClientSubclass(w3, **client_kwargs)
+        return instance
 
     class ConnectionNotEstablished(RuntimeError):
         pass
 
     class SyncTimeout(RuntimeError):
         pass
 
-    def __init__(self, w3, node_technology, version, backend, *args, **kwargs):
-        self.w3 = w3
-        self.node_technology = node_technology
-        self.node_version = version
-        self.backend = backend
-
-        self.log = Logger("blockchain")
-
     @property
     def peers(self):
         raise NotImplementedError
 
     @property
-    def chain_name(self):
+    def chain_name(self) -> str:
         if not self.is_local:
-            return CHAIN_INDEX[self.w3.net.version]
-        return {
-            1337: "geth --dev",
-            5777: "default ganache chain",
-        }.get(self.w3.net.version, "UNKNOWN DEV CHAIN")
+            return PUBLIC_CHAINS[self.w3.net.version]
+        name = LOCAL_CHAINS.get(self.w3.net.version, UNKNOWN_DEVELOPMENT_CHAIN_ID)
+        return name
 
     @property
-    def syncing(self):
+    def syncing(self) -> Union[bool, dict]:
         return self.w3.eth.syncing
 
-    def unlock_account(self, address, password):
-        raise NotImplementedError
-
-    def unlockAccount(self, address, password):
+    def unlock_account(self, address, password) -> bool:
```

Comment:
I like this convention of using underscore_methods to make them obviously distinct from w3 calls